pkg/nsx/services/securitypolicy/store.go

```diff
@@ -1,153 +1,117 @@
+	return *o.Id, nil
 }
 
-func querySecurityPolicy(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors chan error) {
-	defer wg.Done()
-	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeSecurityPolicy) + " AND " + queryTagCondition(service)
-	var cursor *string = nil
-	pageSize := common.PageSize
-	for {
-		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
-		err = common.TransError(err)
-		if _, ok := err.(nsxutil.PageMaxError); ok == true {
-			common.DecrementPageSize(&pageSize)
-			continue
-		}
-		if err != nil {
-			fatalErrors <- err
-		}
-		for _, g := range response.Results {
-			a, err := Converter.ConvertToGolang(g, model.SecurityPolicyBindingType())
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.SecurityPolicy)
-			err2 := service.SecurityPolicyStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+func keyFuncGroup(obj interface{}) (string, error) {
+	o := obj.(model.Group)
+	return *o.Id, nil
+}
+
+func keyFuncRule(obj interface{}) (string, error) {
+	o := obj.(model.Rule)
+	return *o.Id, nil
+}
+
+func indexFuncSecurityPolicy(obj interface{}) ([]string, error) {
+	o := obj.(model.SecurityPolicy)
+	return filterTag(o.Tags), nil
+}
+
+func indexFuncGroup(obj interface{}) ([]string, error) {
+	o := obj.(model.Group)
+	return filterTag(o.Tags), nil
+}
+
+func indexFuncRule(obj interface{}) ([]string, error) {
+	o := obj.(model.Rule)
+	return filterTag(o.Tags), nil
+}
+
+var filterTag = func(v []model.Tag) []string {
+	res := make([]string, 0, 5)
+	for _, tag := range v {
+		if *tag.Scope == key {
+			res = append(res, *tag.Tag)
 		}
 	}
+	return res
 }
 
-func queryRule(service *SecurityPolicyService, wg *sync.WaitGroup, fatalErrors chan error) {
-	defer wg.Done()
-	queryParam := fmt.Sprintf("%s:%s", common.ResourceType, common.ResourceTypeRule) + " AND " + queryTagCondition(service)
-	var cursor *string = nil
-	pageSize := common.PageSize
-	for {
-		response, err := service.NSXClient.QueryClient.List(queryParam, cursor, nil, &pageSize, nil, nil)
-		err = common.TransError(err)
-		if _, ok := err.(nsxutil.PageMaxError); ok == true {
-			common.DecrementPageSize(&pageSize)
-			continue
-		}
-		if err != nil {
-			fatalErrors <- err
-		}
-		for _, g := range response.Results {
-			a, err := Converter.ConvertToGolang(g, model.RuleBindingType())
-			if err != nil {
-				for _, e := range err {
-					fatalErrors <- e
-				}
-			}
-			c, _ := a.(model.Rule)
-			err2 := service.RuleStore.Add(c)
-			if err2 != nil {
-				fatalErrors <- err2
-			}
-		}
-		cursor = response.Cursor
-		if cursor == nil {
-			break
-		}
-		c, _ := strconv.Atoi(*cursor)
-		if int64(c) >= *response.ResultCount {
-			break
+type StoreResource struct {
```

Comment:
how about ResourceStore, SecurityPolicyStore,  RuleStore,  GroupStore?

Reply:
Then it's conflicted with https://github.com/vmware-tanzu/nsx-operator/pull/129/files#diff-53bc6d3b74f0d34b4187204c1bcf7ce4afe93c7f73e60b5de43c40dc44b88af0R47

Reply:
Can we we resolve conflicts, semantically,  ResourceStore, SecurityPolicyStore, RuleStore and GroupStore could be better, In fact, these store is a specific implementation of abstract interface store: 
type Store interface {
	TransResourceToStore(obj *data.StructValue) error
	CRUDResource(obj interface{}) error
}